internal/receiver/discoveryreceiver/evaluator.go

```diff
@@ -0,0 +1,173 @@
+	if !shouldLog || err != nil {
+		return
+	}
+
+	if match.FirstOnly {
+		loggedKey := fmt.Sprintf("%s::%s::%s::%s", endpointID, receiverID.String(), status, matchPattern)
+		if _, ok := e.alreadyLogged.LoadOrStore(loggedKey, struct{}{}); ok {
+			shouldLog = false
+		}
+	}
+
+	return
+}
+
+// correlateResourceAttributes will copy all `from` attributes to `to` in addition to
+// updating embedded base64 config content, if configured, to include the correlated observer ID
+// that is otherwise unavailable to status sources.
+func (e *evaluator) correlateResourceAttributes(from, to pcommon.Map, corr correlation) {
+	receiverType := string(corr.receiverID.Type())
+	receiverName := corr.receiverID.Name()
+
+	observerID := corr.observerID.String()
+	to.InsertString(observerIDAttr, observerID)
+
+	var receiverAttrs map[string]string
+	hasTemporaryReceiverConfigAttr := false
+	receiverAttrs = e.correlations.Attrs(corr.receiverID)
+
+	if e.config.EmbedReceiverConfig {
+		if _, ok := from.Get(receiverConfigAttr); !ok {
+			// statements don't inherit embedded configs in their resource attributes
+			// from the receiver creator, so we should temporarily include it in `from`
+			// so as not to mutate the original while providing the desired receiver config
+			// value set by the initial receiver config parser.
+			from.InsertString(receiverConfigAttr, receiverAttrs[receiverConfigAttr])
+			hasTemporaryReceiverConfigAttr = true
+		}
+	}
+
+	from.Range(func(k string, v pcommon.Value) bool {
+		if k == receiverConfigAttr && e.config.EmbedReceiverConfig {
+			configVal := v.AsString()
+			if updatedConfig, ok := receiverAttrs[receiverUpdatedConfigAttr]; ok {
+				configVal = updatedConfig
+			} else {
+				var err error
+				if updatedConfig, err = addObserverToEncodedConfig(configVal, observerID); err != nil {
+					// log failure and continue with existing config sans observer
+					e.logger.Debug(fmt.Sprintf("failed adding %q to %s", observerID, receiverConfigAttr), zap.String("receiver.type", receiverType), zap.String("receiver.name", receiverName), zap.Error(err))
+				} else {
+					e.logger.Debug("Adding watch_observer to embedded receiver config receiver attrs", zap.String("observer", corr.observerID.String()), zap.String("receiver.type", receiverType), zap.String("receiver.name", receiverName))
+					e.correlations.UpdateAttrs(corr.receiverID, map[string]string{
+						receiverUpdatedConfigAttr: updatedConfig,
+					})
+					configVal = updatedConfig
+				}
+			}
+			v = pcommon.NewValueString(configVal)
+		}
+		to.Insert(k, v)
+		return true
+	})
+	if hasTemporaryReceiverConfigAttr {
+		from.Remove(receiverConfigAttr)
+	}
+}
+
+func addObserverToEncodedConfig(encoded, observerID string) (string, error) {
+	cfg := map[string]any{}
+	dBytes, err := base64.StdEncoding.DecodeString(encoded)
+	if err != nil {
+		return "", err
+	}
+	if err = yaml.Unmarshal(dBytes, &cfg); err != nil {
+		return "", err
+	}
+	cfg["watch_observers"] = []string{observerID}
```

Comment:
What happens if there is already a watch observer? Not sure if that matters.

Reply:
It will be overridden, but the config in question comes from https://github.com/signalfx/splunk-otel-collector/blob/main/internal/receiver/discoveryreceiver/config.go#L229 and this should never happen